addons/hr_holidays/models/hr_holidays.py

```diff
@@ -0,0 +1,416 @@
+
+                meeting = CalendarEvent.with_context(no_email=True).create(meeting_vals)
+                holiday._create_resource_leave()
+                self.write({'meeting_id': meeting.id})
+            elif holiday.holiday_type == 'category':
+                leaves = []
+                for emp in holiday.category_id.employee_ids:
+                    vals = {
+                        'name': holiday.name,
+                        'type': holiday.type,
+                        'holiday_type': 'employee',
+                        'holiday_status_id': holiday.holiday_status_id.id,
+                        'date_from': holiday.date_from,
+                        'date_to': holiday.date_to,
+                        'notes': holiday.notes,
+                        'number_of_days_temp': holiday.number_of_days_temp,
+                        'parent_id': holiday.id,
+                        'employee_id': emp.id
+                    }
+                    leaves.append(self.with_context(mail_notify_force_send=False).create(vals))
+                for leave in leaves:
+                    # TODO is it necessary to interleave the calls?
+                    for sig in ('confirm', 'validate', 'second_validate'):
+                        leave.signal_workflow(sig)
+        return True
+
+    @api.multi
+    def holidays_confirm(self):
+        return self.write({'state': 'confirm'})
+
+    @api.multi
+    def holidays_refuse(self):
+        manager = self.env['hr.employee'].search([('user_id', '=', self.env.uid)], limit=1)
+        for holiday in self:
+            if holiday.state == 'validate1':
+                holiday.write({'state': 'refuse', 'manager_id': manager.id})
+            else:
+                holiday.write({'state': 'refuse', 'manager_id2': manager.id})
+        self.holidays_cancel()
+        return True
+
+    @api.multi
+    def holidays_cancel(self):
+        for holiday in self:
+            # Delete the meeting
+            holiday.meeting_id.unlink()
+
+            # If a category that created several holidays, cancel all related
+            holiday.linked_request_ids.signal_workflow('refuse')
+
+        self._remove_resource_leave()
+        return True
+
+    @api.multi
+    def toggle_payslip_status(self):
+        for holiday in self:
+            holiday.payslip_status = not holiday.payslip_status
+        return True
+
+    @api.multi
+    def _track_subtype(self, init_values):
+        self.ensure_one()
+        if 'state' in init_values and self.state == 'validate':
+            return 'hr_holidays.mt_holidays_approved'
+        elif 'state' in init_values and self.state == 'validate1':
+            return 'hr_holidays.mt_holidays_first_validated'
+        elif 'state' in init_values and self.state == 'confirm':
+            return 'hr_holidays.mt_holidays_confirmed'
+        elif 'state' in init_values and self.state == 'refuse':
+            return 'hr_holidays.mt_holidays_refused'
```

Comment:
Maybe you could have a single `if 'state' in init_values:` and the inside that check dispatch on `self`'s actual state? Also would benefit from a dict mapping state to subtype.